pip/pip-278.md

```diff
@@ -0,0 +1,126 @@
+# Background knowledge
+
+Apache Pulsar is a distributed messaging system that supports multiple messaging protocols and storage methods. Among them, Pulsar Topic Compaction is a mechanism to clean up duplicate messages in topics to reduce storage space and improve system efficiency.
+More topic compaction details can be found in [Pulsar Topic Compaction](https://pulsar.apache.org/docs/en/concepts-topic-compaction/).
+
+# Motivation
+
+Currently, the implementation of Pulsar Topic Compaction is fixed and does not support custom strategy, which limits users from using more Compactor policies in their applications.
+
+For example, we need to parse the Kafka format then compact message in Kop, but the current implementation of Pulsar topic compaction does not support this feature.
+
+Another we also need to support write the compacted data anywhere, S3, in columnar format, or even partitioning.
+
+So we need to make  the whole topic compaction service (include Write API & Read API) pluggable to support more customize compaction service implementation.
+
+# Goals
+
+## In Scope
+
+* Abstract topic compaction service interface and support topic compaction service pluggable.
+
+* Migrate the current implementation to a new interface implementation.
+
+* Makes existing tests compatible with new implementations.
+
+## Out of Scope
+
+For CompactorMetrics, keep the current implementation and don't define related method in topic compaction service interface, in the future will use the `Otel` interface or other mertics API instead.
+
+
+# High Level Design
+
+To make the whole topic compaction service pluggable, we need to abstract `TopicCompactionService` interface, it can provide the capability that the compactor has and provide the read API to read entries from compacted data.
```

Comment:
Would you mind putting an Inheritance chart to help others understand this part? :)

Reply:
Yeah, I will add a class diagram